packages/driver/cypress/integration/commands/assertions_spec.js

```diff
@@ -1,6 +1,39 @@
 const { assertLogLength } = require('../../support/utils')
 const { $, _ } = Cypress
 
+const captureCommands = () => {
```

Comment:
Do we not manage / track the number of snapshots taken during a test? I assumed this data was reported to the dashboard?

Reply:
These DOM snapshots are the ones visible when you hover over test steps and it displays how the application looked at that point in time. They're very ephemeral - we only take them in open mode, and they live in the browser running tests, not even sent to the cypress server. We discard the snapshots regularly (if the number of tests run is greater than `numTestsKeptInMemory`, whenever we rerun tests, etc).

We didn't even have an internal event for counting them until I added it in this PR (`Cypress.action('cy:snapshot', name)`) so this utility could listen for it.

Reply:
> Could be a bit off here on the flow so if my comments seem way off, sorry in advance! I have a few questions....
> 
>     1. Mind adding a few more details on how override the assertion method is fixing this? It's not 100% apparent.

Yeah, the key here is that I'm not doing something new, I'm *covering another case* and doing the same thing we were already doing in the other location.

`verifyUpcomingAssertions` has two exit points. It can return on either line 328 if there are no upcoming commands:
```
if (!cmds.length) {
        return Promise
        .try(ensureExistence)
        .then(onPassFn)
        .catch(onFailFn)
}
```

or it can return on line 490:

```
     cy.state('overrideAssert', overrideAssert)

      return Promise
      .reduce(fns, assertions, [subject])
      .then(() => {
        restore()
        ...
      }).catch(() => {
        restore()
        ...
     })
```

When there *are* upcoming commands in the queue - when we're verifying assertions like ".should('have.class', 'foo')", then we override the assert function in a way that passes in `verifying: true` - line 442 inside `overrideAssert`. This PR makes the same thing happen for implicit assertions.

So to try and put it in English a bit better: In general, when an assertion fails, we want to take a snapshot, so we can show the users what went wrong. But when we're when we're in the middle of `cy.get()` checking to see if we should move on to following commands or sit here retrying cy.get() - that is to say we're *verifyingUpcomingAssertions* - then we execute the assertion code every 50ms, but we're only doing to to *see if we're ready to move on* - not in order to pass or fail the assertion itself!

That's why what `verifying = false | true` on line 128 in `assertFn` represents. If it's false, we are actually executing an assertion - we do things like put an error on the log message and take a snapshot. If we're only verifying upcoming assertions, then we still execute the assertion code - but verifying = true tells us not to do things that we only want to do once at the very end (eg, take a snapshot to display to the user).

>     2. > Ensure that we don't snapshot for assertions that fail our retry logic already takes a snapshot at the end when a test actually fails.
> 
> 
> Is this true when retries are set to 0 in the configuration? Does your fix handle this scenario?

I'll update the PR description here - I wrote it based on an earlier misunderstanding of the bug. It's what the first test I added, with `timeout: 0` is demonstrating - we were taking extra snapshots *unrelated to retries.*

Retries are not the cause of the bug, they just make it way more obvious. Setting {timeout: 0} in that first test ensures that no retries occur - but if you run that test against develop (or comment out the code additions in this branch), you'll see that three snapshots were occurring even though there were 0 retries.

>     2. Did you consider updating [the retries logic here](https://github.com/cypress-io/cypress/blob/issue-18549-no-unnecessary-snapshotting/packages/driver/src/cy/retries.ts?rgh-link-date=2021-12-16T21%3A42%3A37Z#L57-L65) to account for this given the [finishAssertions callback](https://github.com/cypress-io/cypress/blob/issue-18549-no-unnecessary-snapshotting/packages/driver/src/cypress/cy.ts?rgh-link-date=2021-12-16T21%3A42%3A37Z#L250-L254) provided to the retries class which uses the [assertion class' finishAssertion method which also takes the snapshot](https://github.com/cypress-io/cypress/blob/issue-18549-no-unnecessary-snapshotting/packages/driver/src/cy/assertions.ts?rgh-link-date=2021-12-16T21%3A42%3A37Z#L205)?

Yeah, my initial attempt to fix this was in the retries logic - but that's why the first test is so important. It demonstrates the problem outside the context of retries. If we're triple snapshoting without retries, then making sure we only do three snapshots is still not the correct solution - there should only be one!

I really should have updated the description before sending this over.